[DR-3182] Use billing profile in Azure snapshot export without checking caller's access

okotsopoulos opened the pull request:
https://broadworkbench.atlassian.net/browse/DR-3182

Adam Mullen was getting an "unauthorized" error when trying to initiate the first step of an Azure snapshot export (TDR Dev job 0-Aks54WQR2eM7-uRPBI_Q)  The issue is that while he's a snapshot steward, he has no permissions on the billing profile directly.

I believe that this check is done in oversight and is not necessary.  I do not see a similar check performed for GCP snapshot export.  At this point we've already verified that the caller holds the needed action on the snapshot.

In addition to changing the billing profile fetch to an unchecked method, I expanded unit testing to verify our snapshot authorization check and improved the error thrown when trying to fetch an inaccessible billing profile.

okotsopoulos committed:
Throw more descriptive error when unauthorized to fetch billing profile

This new message aligns with the errors we throw when handling Sam unauthorization errors.

okotsopoulos committed:
Use billing profile in Azure snapshot export without checking caller's access

This matches the behavior when exporting GCP snapshots.
Added unit tests to confirm that we check that the caller holds the right action on snapshots when exporting.

pshapiro4broad commented on src/test/java/bio/terra/app/controller/SnapshotsApiControllerTest.java:
```diff
@@ -0,0 +1,130 @@
+
+@ActiveProfiles({"google", "unittest"})
+@ContextConfiguration(classes = {SnapshotsApiController.class, GlobalExceptionHandler.class})
+@Tag("bio.terra.common.category.Unit")
```
This can use the `Unit.TAG` field
```suggestion
@Tag(Unit.TAG)
```

pshapiro4broad commented on src/test/java/bio/terra/app/controller/SnapshotsApiControllerTest.java:
```diff
@@ -0,0 +1,130 @@
+  }
+
+  /** Mock so that the user does not hold `iamAction` on the dataset. */
+  private void mockForbidden(IamAction iamAction) {
```
Not a big deal but it looks like `iamAction` is always `IamAction.EXPORT_SNAPSHOT` so it could be inlined in these two helper methods.

okotsopoulos replied:
I ended up inlining the `mockForbidden` method as it was only called once, but I elected to pass in the `IamAction` to the other helper method since it's nice to be able to reference it in both the mock set-up and later verification.

pshapiro4broad commented on src/main/java/bio/terra/service/profile/ProfileService.java:
```diff
@@ -185,7 +185,8 @@ public EnumerateBillingProfileModel enumerateProfiles(
     if (!iamService.hasAnyActions(user, IamResourceType.SPEND_PROFILE, id.toString())) {
-      throw new IamUnauthorizedException("unauthorized");
+      throw new IamUnauthorizedException(
+          "User '" + user.getEmail() + "' does not have access to billing profile '" + id + "'.");
```
Do we usually include the user email in IAM error messages? I don't think it can expose any information but the email is always the caller's email so it seems like they'd already have that information.

okotsopoulos replied:
We do include this!  I've also had to help users debug authorization issues that were caused by them not logging in as the expected account.

pshapiro4broad commented on src/main/java/bio/terra/service/snapshot/flight/export/SnapshotExportWriteManifestAzureStep.java:
```diff
@@ -67,7 +67,10 @@ public StepResult doStep(FlightContext context) throws InterruptedException, Ret
-    BillingProfileModel billingProfile = profileService.getProfileById(billingProfileId, userReq);
+    // We do not check that the caller has direct access to the billing profile:
+    // It is sufficient that they hold the necessary action on the snapshot, which is checked
+    // before calling the flight.
```
Is there a case where we'd want to re-check a permission inside a flight, to make sure that the permission hasn't changed since the flight was created? Or maybe this is isn't an interesting/common enough case to bother handling.

okotsopoulos replied:
No, I don't think there's a case for this.  When initiating flights in general, we check at the start that the user has the permission to launch it.

pshapiro4broad commented on src/test/java/bio/terra/app/controller/SnapshotsApiControllerTest.java:
```diff
@@ -0,0 +1,130 @@
+                    String.valueOf(VALIDATE_PRIMARY_KEY_UNIQUENESS)))
+        .andExpect(status().isForbidden());
+
+    verifyAuthorizationCall(iamAction);
```
I get why this test is verifying the call, but I wonder if it's necessary. If the IAM check was never made, how could the API call return forbidden? Or maybe this is a just a good convention to follow in any case without thinking too much about what the code is actually doing.

okotsopoulos replied:
Good questions.  I don't think it's strictly necessary but I do like the convention myself.

pshapiro4broad reviewed (approved):
Looks good to me

nmalfroy reviewed (approved):
thanks for fixing this!

okotsopoulos committed:
Sonar code smell: JUnit5 test class needn't be public

JUnit5 is more tolerant regarding the visibilities of Test classes than JUnit4, which required everything to be public.

In this context, JUnit5 test classes can have any visibility but private, however, it is recommended to use the default package visibility, which improves readability of code.

okotsopoulos committed:
@pshapiro4broad PR feedback

- Use Unit.TAG on unit test class rather than hardcoded path
- Inline test helper method only used once
